azure.js

```diff
@@ -44,37 +46,117 @@ module.exports = (args, sendTo) => {
   }
 
   const processVirtualMachineScaleSet = function (client, ips, localIps, vmssName) {
-    client.networkInterfaces._listVirtualMachineScaleSetNetworkInterfaces(args.azure, vmssName, function(err, results) {
-      if (err) {
-        console.log(err);
-        return;
-      }
-
-      results.forEach(result => {
-        result.ipConfigurations.forEach(ip => {
-          if (!localIps.includes(ip.privateIPAddress)) {
-            ips.push(ip.privateIPAddress)
+    client.networkInterfaces
+      ._listVirtualMachineScaleSetNetworkInterfaces(args.azure, vmssName,
+        function (err, results) {
+          if (err) {
+            console.log(err);
+            return;
           }
+
+          results.forEach(result => {
+            result.ipConfigurations.forEach(ip => {
+              if (!localIps.includes(ip.privateIPAddress)) {
+                ips.push(ip.privateIPAddress)
+              }
+            })
+          })
+
+          updateSendTo(ips)
         })
-      })
+  }
 
-      updateSendTo(ips)
+  const processVirtualMachineScaleSetPromise = function (
+    client,
+    vmssName
+  ) {
+    return new Promise((resolve, reject) => {
+      client.networkInterfaces._listVirtualMachineScaleSetNetworkInterfaces(
+        args.azure,
+        vmssName,
+        function (err, results) {
+          if (err) {
+            // Do nothing if the VMSS is deleted, so that for next deployment inter VMSS forwarding will work
+            if (err.code !== 'ParentResourceNotFound') {
+              console.log(err)
+            }
+            resolve(null)
+          }
+          resolve(results)
+        }
+      )
     })
   }
 
+  async function getVMDetailsByTag(tagKey, tagValue) {
+
+    const credential = new DefaultAzureCredential();
+
+    const client = new ComputeManagementClient(credential, subscriptionId);
+
+    // Get the list of VMs
+    //const vms = await client.virtualMachines.listAll().byPage().next();
+
+    const vmss = await client.virtualMachineScaleSets.listAll().byPage().next();
+
+    const filteredVMSSs = vmss.value
+      .filter(vms => vms.tags && vms.tags[tagKey] === tagValue)
+      .map(vms => vms.name);
+
+    return filteredVMSSs;
+
+  }
+
+  const processMultipleVirtualMachineScaleSets = function (
+    client,
+    ips,
+    localIps,
+    tagKey,
+    tagValue
+  ) {
+    getVMDetailsByTag(tagKey, tagValue).then(filteredVMSS => {
+      const requests = filteredVMSS.map((vmssName) => {
+        return new Promise(async resolve => {
+          const results = await processVirtualMachineScaleSetPromise(client, vmssName)
+          if (results && results.length) {
+            results.forEach(result => {
+              result.ipConfigurations.forEach(ip => {
+                if (!localIps.includes(ip.privateIPAddress)) {
+                  ips.push(ip.privateIPAddress)
+                }
+              })
+            })
+            resolve(ips)
```

Comment:
After this line I think there should be a `return` or an `else` so control doesn't reach `resolve([])` 2 lines below

Reply:
Return is not required here we are using outer scoped variable ips, we need this promise just for retrieving all VM's IPs at once.

Reply:
After the first resolve() is called, the second call doesn't do anything.

Reply:
Resolved